### PR TITLE
Hipp 265

### PR DESCRIPTION
**Pull request by tigdik**
Update backend to change status of a scope to APPROVED

### CODE REVIEW

**Comment by laurahk** [comment]
Some warnings about unused imports:
```
[warn] /Users/hmrc/Documents/projects/api-hub-applications/app/uk/gov/hmrc/apihubapplications/repositories/ApplicationsRepository.scala:21:31: Unused import
[warn] import org.mongodb.scala.bson.Document
[warn]                               ^
[warn] /Users/hmrc/Documents/projects/api-hub-applications/app/uk/gov/hmrc/apihubapplications/repositories/ApplicationsRepository.scala:33:34: Unused import
[warn] import scala.concurrent.duration._
[warn]                                  ^
[warn] /Users/hmrc/Documents/projects/api-hub-applications/app/uk/gov/hmrc/apihubapplications/repositories/ApplicationsRepository.scala:34:26: Unused import
[warn] import scala.concurrent.{Await, ExecutionContext, Future}
[warn]                          ^
[warn] three warnings found
```

File: test/uk/gov/hmrc/apihubapplications/controllers/ApplicationsControllerSpec.scala [hunk 1]
```diff
@@ -253,8 +254,76 @@ class ApplicationsControllerSpec
...
+          .withBody(json)
+
+        val result = route(fixture.application, request).value
+        status(result) mustBe NoContent.code
+
+        verify(fixture.repository).setScope(ArgumentMatchers.eq(appId), ArgumentMatchers.eq(envName), ArgumentMatchers.eq(scopeName), ArgumentMatchers.eq(updateScope))
+      }
+    }}
+
+    "must return 404 Not Found when trying to set scope on the application that does not exist in DB" in {
```
Comment (laurahk) [inline test/uk/gov/hmrc/apihubapplications/controllers/ApplicationsControllerSpec.scala:281]: These tests are not nested inside "set scope to APPROVED" properly:
TestNesting

File: test/uk/gov/hmrc/apihubapplications/repositories/ApplicationsRepositorySpec.scala [hunk 2]
```diff
@@ -85,6 +86,23 @@ class ApplicationsRepositorySpec
       (result \ 1 \"environments" \ 0) mustBe JsDefined(JsString("prod"))
     }
 
+    "must successfully de-serialise a collection of new scopes" in {
```
Comment (laurahk) [inline test/uk/gov/hmrc/apihubapplications/repositories/ApplicationsRepositorySpec.scala:89]: Thanks for adding this missing test 👍

File: app/uk/gov/hmrc/apihubapplications/repositories/ApplicationsRepository.scala [hunk 3]
```diff
@@ -38,7 +41,10 @@ class ApplicationsRepository @Inject()
     collectionName = "applications",
     mongoComponent = mongoComponent,
     domainFormat   = mongoApplicationFormat,
-    extraCodecs = Seq(Codecs.playFormatCodec(Scope.scopeFormat)),
+    extraCodecs = Seq(Codecs.playFormatCodec(Scope.scopeFormat),
+      Codecs.playFormatCodec(UpdateScopeStatus.updateScopeStatusFormat),
+      Codecs.playFormatCodec(UpdateScopeStatus.updateScopeStatusFormat)
```
Comment (laurahk) [inline app/uk/gov/hmrc/apihubapplications/repositories/ApplicationsRepository.scala:46]: duplicate line here

File: app/uk/gov/hmrc/apihubapplications/repositories/ApplicationsRepository.scala [hunk 4]
```diff
@@ -86,6 +92,21 @@ class ApplicationsRepository @Inject()
     }
   }
 
+  def setScope(applicationId:String, env:String, scope:String, updateStatus:UpdateScopeStatus)= {
+    stringToObjectId(applicationId) match {
+      case Some(appIdObject) =>
+          val updates = Seq(
+            set(f"environments.$env.scopes.$$.status", updateStatus.status.toString),
```
Comment (laurahk) [inline app/uk/gov/hmrc/apihubapplications/repositories/ApplicationsRepository.scala:99]: `updateStatus.status.toString`
Is `.toString` the correct way to get the name from an enumeration? Is it not something like `.value`?

File: it/uk/gov/hmrc/apihubapplications/ApplicationsRepositoryIntegrationSpec.scala [hunk 5]
```diff
@@ -193,4 +194,35 @@ class ApplicationsRepositoryIntegrationSpec
     }
   }
 
+  "update scope to APPROVED" - {
+    "must return true if in scope status was updated successfully to APPROVED" in {
```
Comment (laurahk) [inline it/uk/gov/hmrc/apihubapplications/ApplicationsRepositoryIntegrationSpec.scala:198]: Is this supposed to say "the" instead of "in"?

Comment (laurahk) [reply it/uk/gov/hmrc/apihubapplications/ApplicationsRepositoryIntegrationSpec.scala:198]: The test assets the correct behaviour (that Pending is changed to Approved) but this doesn't match the test description - you don't actually check in the test that `true` is returned.

Comment (laurahk) [reply it/uk/gov/hmrc/apihubapplications/ApplicationsRepositoryIntegrationSpec.scala:198]: And it should say "Some(true)"

File: conf/app.routes [hunk 6]
```diff
@@ -1,8 +1,10 @@
...
-POST       /applications/:id/environments/scopes        uk.gov.hmrc.apihubapplications.controllers.ApplicationsController.addScopes(id: String)
+POST       /applications                                                            uk.gov.hmrc.apihubapplications.controllers.ApplicationsController.registerApplication
+POST       /applications/:id/environments/scopes                                    uk.gov.hmrc.apihubapplications.controllers.ApplicationsController.addScopes(id: String)
 
-GET        /applications/pending-scopes                 uk.gov.hmrc.apihubapplications.controllers.ApplicationsController.pendingScopes
-GET        /applications                                uk.gov.hmrc.apihubapplications.controllers.ApplicationsController.getApplications
-GET        /applications/:id                            uk.gov.hmrc.apihubapplications.controllers.ApplicationsController.getApplication(id: String)
+PUT        /applications/:id/environment/:environment/scopes/:scopename            uk.gov.hmrc.apihubapplications.controllers.ApplicationsController.setApprovedScope(id:String,environment:String, scopename:String)
```
Comment (laurahk) [inline conf/app.routes:6]: `environment` should be `environments` (with an s) according to the ticket. When you change this here you will have to update the tests that call the endpoint.

**Comment by laurahk** [comment]
@tigdik You also had a conversation with me last week saying you were going to change the structure of the code such that you would read the application, change the status, and then write the whole application back to the database. And that this would introduce another layer in the code. It doesn't look like you've done that here. What happened to that idea?

File: app/uk/gov/hmrc/apihubapplications/repositories/ApplicationsRepository.scala (hunk 4 above)
Comment (tigdik) [reply app/uk/gov/hmrc/apihubapplications/repositories/ApplicationsRepository.scala:97]: status is not Enumeration

File: app/uk/gov/hmrc/apihubapplications/repositories/ApplicationsRepository.scala [hunk 7]
```diff
@@ -86,6 +92,21 @@ class ApplicationsRepository @Inject()
...
+            set("lastUpdated", LocalDateTime.now().toString)
```
[the remark this answers is not in the document]
Comment (laurahk) [reply app/uk/gov/hmrc/apihubapplications/repositories/ApplicationsRepository.scala:97]: Talked to Vlad about this and setting the lastUpdated time and in a previous ticket, when we add a new scope we also update the lastUpdated time. Going to leave this in and talk to Whitney about the requirements of it.

File: it/uk/gov/hmrc/apihubapplications/ApplicationsIntegrationSpec.scala [hunk 8]
```diff
@@ -45,6 +46,7 @@ class ApplicationsIntegrationSpec
 
   private val wsClient = app.injector.instanceOf[WSClient]
   private val baseUrl  = s"http://localhost:$port"
+  //private val baseUrl  = s"http://localhost:9000"
```
Comment (laurahk) [inline it/uk/gov/hmrc/apihubapplications/ApplicationsIntegrationSpec.scala:49]: please remove this comment

Comment (tigdik) [reply it/uk/gov/hmrc/apihubapplications/ApplicationsIntegrationSpec.scala]: done

**Review by laurahk — APPROVED** [review]

**Pull request merged**